cwltool/validate_js.py

```diff
@@ -0,0 +1,179 @@
+
+from . import process
+from .command_line_tool import CommandLineTool
+from .expression import scanner as scan_expression
+from .process import Process
+from .sandboxjs import (JavascriptException, code_fragment_to_js,
+                        exec_js_process, execjs)
+
+_logger = logging.getLogger("cwltool")
+
+def is_expression(tool, schema):
+    # type: (Union[CommentedMap, Any], avro.schema.Schema) -> bool
+    return isinstance(schema, avro.schema.EnumSchema) and schema.name == "Expression" and isinstance(tool, str)
+
+def get_expressions(tool, schema, source_line=None):
+    # type: (Union[CommentedMap, Any], avro.schema.Schema, SourceLine) -> List[Tuple[str, SourceLine]]
+    if is_expression(tool, schema):
+        return [(tool, source_line)]
+    elif isinstance(schema, avro.schema.UnionSchema):
+        valid_schema = None
+
+        for possible_schema in schema.schemas:
+            if is_expression(tool, possible_schema):
+                return [(tool, source_line)]
+            elif validate_ex(possible_schema, tool, raise_ex=False):
+                valid_schema = possible_schema
+
+        return get_expressions(tool, valid_schema, source_line)
+    elif isinstance(schema, avro.schema.ArraySchema):
+        if not isinstance(tool, list):
+            return []
+
+        return list(itertools.chain(*
+            map(lambda x: get_expressions(x[1], schema.items, SourceLine(tool, x[0])), enumerate(tool))
+        ))
+    elif isinstance(schema, avro.schema.RecordSchema):
+        if not isinstance(tool, Dict):
+            return []
+
+        expression_nodes = []
+
+        for schema_field in schema.fields:
+            if schema_field.name in tool:
+                expression_nodes.extend(get_expressions(
+                    tool[schema_field.name],
+                    schema_field.type,
+                    SourceLine(tool, schema_field.name)
+                ))
+
+        return expression_nodes
+    else:
+        return []
+
+def should_include_jshint_message(error_code):
+    # type: (str) -> bool
+    include_warnings = [
+        "W117", # <VARIABLE> not defined
+        "W104", "W119"  # using ES6 features
+    ]
+
+    return error_code[0] == "E" or error_code in include_warnings
+
+JSHintJSReturn = namedtuple("jshint_return", ["errors", "globals"])
+
+def jshint_js(js_text, globals=None, options=None):
+    # type: (str, List[str], Dict) -> Tuple[List[str], List[str]]
+    if globals is None:
+        globals = []
+    if options is None:
+        options = {
+            "includewarnings": [
+                "W117", # <VARIABLE> not defined
+                "W104", "W119"  # using ES6 features
+            ],
+            "strict": "implied",
+            "esversion": 5
+        }
+
+    linter_folder = resource_filename(__name__, "jshint")
+
+    returncode, stdout, stderr = exec_js_process(
+        path.join(linter_folder, "jshint_wrapper.js"),
```

Comment:
Here you should use get_data from https://github.com/common-workflow-language/cwltool/blob/master/tests/util.py (maybe that should be moved to the cwltool package?)

This is why the conformance tests are failing 🙂